docs: Add Port Range Information

nathanjsweet opened the pull request:
~- SCTP and Port Ranges are now supported, both are removed from the unsupported features table.~
- DNS rules and L7 rules do not support port ranges yet. Notes are added to call attention to this.
- Add a Port Range example.

**Edit:**
- SCTP support will be addressed in a different PR.

joestringer commented on Documentation/network/kubernetes/policy.rst:
```diff
@@ -50,10 +50,6 @@ Known missing features for Kubernetes Network Policy:
 +===============================+===================+
 | ``ipBlock`` set with a pod IP | :gh-issue:`9209`  |
 +-------------------------------+-------------------+
-| SCTP                          | :gh-issue:`5719`  |
```
If SCTP is now supported, we should probably consider adding SCTP to the upstream k8s suite in `test/kubernetes-test.sh`. It looks like it's skipped for now.

joestringer replied:
Logically it'd be nice to process the SCTP change separately to ensure proper scrutiny. If it's really supported on older versions then this could also be a candidate for fixing the docs in older branches.

joestringer commented on Documentation/security/policy/language.rst:
```diff
@@ -719,11 +721,15 @@ which is defined as follows:
-                // parsed as a single uint16. In the future, this field may support
-                // ranges in the form "1024-2048"
+                // Port can be an L4 port number, or a name in the form of "http"
+                // or "http-8080".
```
Is this also new, supporting arbitrary named ports? Do we have tests for this too?

nathanjsweet replied:
>Is this also new, supporting arbitrary named ports? Do we have tests for this too?

No, this is not new. This is a fragment of the previous comment. The only change in this comment is removing the conjecture about supporting port ranges as a hyphenated string (e.g. "8080-8090").

joestringer commented on examples/policies/l4/l4_port_range.yaml:
```diff
@@ -0,0 +1,14 @@
+    - toPorts:
+      - ports:
+        - port: "80"
+          endPort: 444
```
The API type described above is a string, shouldn't this be the following?

```suggestion
          endPort: "444"
```

joestringer replied:
(It's actually an int32 in the API)

joestringer commented on Documentation/security/policy/language.rst:
```diff
@@ -719,11 +721,15 @@ which is defined as follows:
+                // or "http-8080".
                 Port string `json:"port"`
 
+                // EndPort can only be an L4 port number.
```
Interesting corner case - is it OK for a user to specify something like this?
```
port: "http"
endPort: "8080"
```

nathanjsweet replied:
No. I will be expanding this comment.

joestringer commented on Documentation/security/policy/language.rst:
```diff
@@ -719,11 +721,15 @@ which is defined as follows:
+                // EndPort can only be an L4 port number.
+                //
+                // +optional
+                EndPort string `json:"endPort,omitempty"`
```
Hmm. The type here is incorrect:

https://github.com/cilium/cilium/commit/4fe60e5214347

joestringer replied:
^ Still inconsistent, the actual API is int32

nathanjsweet replied:
:facepalm:

joestringer commented on Documentation/security/policy/language.rst:
```diff
@@ -719,11 +721,15 @@ which is defined as follows:
         // PortProtocol specifies an L4 port with an optional transport protocol
         type PortProtocol struct {
-                // Port is an L4 port number. For now the string will be strictly
-                // parsed as a single uint16. In the future, this field may support
-                // ranges in the form "1024-2048"
```
I guess we excluded this option in order to be more consistent with k8s APIs?

nathanjsweet replied:
Yes, and it was very difficult to implement.

nathanjsweet committed:
docs: Add Port Range Information

    - Port Ranges are now supported so it is removed from
    the unsupported features table.
    - DNS rules and L7 rules do not support port ranges yet.
    Notes are added to call attention to this.
    - Add a Port Range example.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

joestringer commented on Documentation/security/policy/language.rst:
```diff
@@ -753,6 +760,29 @@ only be able to emit packets using TCP on port 80, to any layer 3 destination:
+
+        .. literalinclude:: ../../../examples/policies/l4/l4_port_range.json
+
+.. note::  "endPort" is ignored when when "port" is a named port.
```
A few points here:
- We already declared this above. Do we need to write it twice?
- In context, this is written as a callout immediately after an unrelated example. Should it be elsewhere?
- Does this need to be a callout? I would argue that unless there's something particularly notable about this, it should just be plain text and not inside a `.. note::` stanza. For instance if it will catch out a significant number of people unless we call it out then it may make sense. I doubt that named ports are so frequently used or are likely to be used in conjunction with port ranges.
```suggestion
```

joestringer commented on Documentation/security/policy/language.rst:
```diff
@@ -929,6 +959,8 @@ latter rule will have no effect.
           an *HTTP 403 access denied* is sent back for HTTP requests which
           violate the policy, or a *DNS REFUSED* response for DNS requests.
 
+.. note:: Layer 7 rules do not currently support port ranges.
```
Hmm, this section is also getting a bit out-of-control with the `..note::` usage. If there are several notes together, then we are not emphasizing any particular one of them, and that means none of them should be notes.

We can follow up on this separately though, this change is not making things much worse.

joestringer commented on Documentation/security/policy/language.rst:
```diff
@@ -659,6 +659,8 @@ port. When the application makes a request for my-remote-service.com, Cilium
 the ``toFQDNs.matchName`` rule.
 
+.. note::  DNS rules do not currently support port ranges.
+
```
Again in context, I think the note here calls out information that is not very relevant to what the reader is processing. The previous paragraph is talking about the example that comes immediately after this note, then there's the example after this, and in the middle is this note which is talking about something else.

DNS rules are L7 rules, and the note is under the L7 rules section as well, so we can probably drop this. If you feel we are not communicating the constraints around port ranges well enough, then I think we should collect all of these notes and put them as plain text under the port ranges section so that any user who attempts to use port ranges can look at the port ranges section and immediately understand how to use the feature and how not to use the feature.
```suggestion
```

joestringer reviewed (changes requested):
I think we should consolidate the restrictions around port ranges into the section where port ranges are introduced and described, and also avoid using notes so much (see [this guidance](https://docs.cilium.io/en/stable/contributing/docs/docsstyle/#callouts) in the docs).

tommyp1ckles commented on Documentation/security/policy/language.rst:
```diff
@@ -719,11 +719,16 @@ which is defined as follows:
-                // parsed as a single uint16. In the future, this field may support
-                // ranges in the form "1024-2048"
+                // Port can be an L4 port number, or a name in the form of "http"
+                // or "http-8080". EndPort is ignored if Port is a named port.
```
Non blocking question:  Do we really want to continuing maintaining a mirror of this k8s API struct in docs, feels like it will be easy for this to get out of date easily.

joestringer replied:
I guess it's a balance. We don't change the API super often. Those structs are ultimately the source of truth for how the API behaves though. Perhaps in an ideal world we'd have docs tooling to automatically embed the latest API struct definitions in here rather than manually updating them.

Other than that, we can either copy the text over into the docs or we can copy the text plus definitions over, either way we have the question of how to make sure the docs describe the semantics of the API. I don't know if it makes a significant difference between the two.
